beetsplug/tidal.py

```diff
@@ -0,0 +1,863 @@
+        except tidalapi.exceptions.ObjectNotFound:
+            self._log.debug(f"No track for ID {tidal_track_id}")
+            return None
+
+        return self._track_to_trackinfo(track, track.album)
+
+    def candidates(self, items, artist, album, va_likely, extra_tags):
+        """Returns TIDAL album candidates for a specific set of items"""
+        candidates = []
+
+        self._log.debug(
+            "Searching for candidates using tidal_album_id from items"
+        )
+        for item in items:
+            if item.get("tidal_album_id", None):
+                try:
+                    albumi = self._album_to_albuminfo(
+                        self.sess.album(item.tidal_album_id)
+                    )
+                    candidates.append(albumi)
+                except tidalapi.exceptions.ObjectNotFound:
+                    self._log.debug(
+                        f"No album found for ID {item.tidal_album_id}"
+                    )
+
+        self._log.debug(
+            f"{len(candidates)} Candidates found using tidal_album_id from items!"
+        )
+        self._log.debug("Searching for candidates using _search_album search")
+
+        if va_likely:
+            candidates += [
+                self._album_to_albuminfo(x)
+                for x in self._search_album(
+                    album, limit=self.config["metadata_search_limit"].get(int)
+                )
+            ]
+        else:
+            candidates += [
+                self._album_to_albuminfo(x)
+                for x in self._search_album(
+                    f"{artist} {album}",
+                    limit=self.config["metadata_search_limit"].get(int),
+                )
+            ]
+
+        self._log.debug(f"_get_lyrics cache: {self._get_lyrics.cache_info()}")
+        self._log.debug(
+            f"_tidal_search cache: {self._tidal_search.cache_info()}"
+        )
+
+        return candidates
+
+    def item_candidates(self, item, artist, album):
+        """Returns TIDAL track candidates for a specific item"""
+        self._log.debug(f"Searching TIDAL for {item}!")
+
+        return (
+            self._search_from_metadata(
+                item, limit=self.config["metadata_search_limit"].as_number()
+            )
+            or []
+        )
+
+    def _album_to_albuminfo(self, album):
+        """Converts a TIDAL album to a beets AlbumInfo
+
+        :param album: An album obtained from the TIDAL API
+        :type album: tidalapi.media.Album
+        :return: A beets AlbumInfo type created with the provided data
+        :rtype: beets.autotag.hooks.AlbumInfo
+        """
+        tracks = []
+
+        # Process tracks
+        # Not using sparse albums as we already have the album
+        # so it's not using up any additional API calls.
+        for track in album.tracks(sparse_album=False):
+            tracks.append(self._track_to_trackinfo(track, album))
+
+        albuminfo = AlbumInfo(
```

Comment:
there are a couple more attributes that should be captured - `label` under `album.copyright`, `cover_art_url` under `album.image(1280)`

Reply:
Alright, I've added that in.

album.copyright tends to have the copyright year in the label, is that something I need to regex/strip out or is it fine as is?

Reply:
Just need the label information.

Reply:
Oh ok, I just committed that.

I made it a regex as TIDAL has no formatting standard for the copyright, and it seems to work on the albums I've tested it with thus far